### PR TITLE
[Feature] BoundedExpression, BoundedBinaryExpression, and BoundedUnaryExpression

### PR DESCRIPTION
**Pull request by matthew-mccall**
This adds `BoundedExpression` an Expression which has a lower bound and an upper bound. It is meant for use cases such as definite integrals (`BoundedBinaryExpression`), and summations (`BoundedUnaryExpression`).

### CODE REVIEW

File: include/Oasis/DefiniteIntegral.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+#ifndef DEFINITEINTEGRAL_HPP
+#define DEFINITEINTEGRAL_HPP
+
+#include "TernaryExpression.hpp"
+
+namespace Oasis {
+
+template <IExpression OperandT, IExpression LowerBoundT, IExpression UpperBoundT>
+class DefiniteIntegral : public TernaryExpression<DefiniteIntegral, OperandT, LowerBoundT, UpperBoundT> {
```
Comment (404Naz) [inline include/Oasis/DefiniteIntegral.hpp:13]: What would happen if OperandT had multiple variables? How does it decide which variable to use?

**Review by 404Naz — COMMENTED** [review]
Read Comment.

File: include/Oasis/DefiniteIntegral.hpp (hunk 1 above)
Comment (matthew-mccall) [reply include/Oasis/DefiniteIntegral.hpp:13]: You're right, in the case of DefiniteIntegral, the operand should be an Integral such that the variable of Integration can be determined

File: include/Oasis/Summation.hpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+#ifndef OASIS_SUMMATION_HPP
+#define OASIS_SUMMATION_HPP
+
+#include "BoundedUnaryExpression.hpp"
+
+namespace Oasis {
+
+template <IExpression OperandT, IExpression LowerBoundT, IExpression UpperBoundT>
+class Summation : public BoundedUnaryExpression<Summation, OperandT, LowerBoundT, UpperBoundT> {
```
Comment (404Naz) [inline include/Oasis/Summation.hpp:13]: Should have a variable in case of nesting or multiple variables.

**Review by 404Naz — COMMENTED** [review]
Summation should have a reference variable so it knows if it should be treating the variable as a real number or as a counter/iterative variable.

**Review by 404Naz — APPROVED** [review]
Looks good to me

**Pull request merged**